code/dynamic_programming/src/digit_dp/digit_dp.py

```diff
@@ -0,0 +1,75 @@
+# Part of Cosmos by OpenGenus Foundation
+# By: Shamama Zehra (https://github.com/shamamaz)
+
+# How many numbers x are there in the range a to b
+# where the digit d occurs exactly k times in x?
+# dp[pos][count][f] = Number of valid numbers <= b from this state
+# pos = current position from left side (zero based)
+# count = number of times we have placed the digit d so far
+# f = the prefix of the number we are building is smaller than prefix in b [0 = no, 1 = yes]
+import numpy as np
+
+def digit_dp(num, pos, count, f):
+    global dp
```

Comment:
Do you _really_ need globals? Why not just pass values around as parameters?

Reply:
With this one, because this function is called recursively and edits the same array, using globals is a fairly direct approach. The alternative would be to pass in the dp array to the function each time which seems memory intensive